Empire/Empire.cpp

```diff
@@ -3372,9 +3372,11 @@ void Empire::ResetMeters() {
 }
 
 void Empire::UpdateOwnedObjectCounters() {
-    // ships of each species and design
+    // ships of each species and design, parts by type and class
     m_species_ships_owned.clear();
     m_ship_designs_owned.clear();
+    m_ship_parts_owned.clear();
+    m_ship_part_class_owned.clear();
     for (std::map<int, TemporaryPtr<UniverseObject> >::iterator ship_it = Objects().ExistingShipsBegin();
```

Comment:
Probably doesn't matter for a once-per-turn calculation, but could calculate number of parts in each design once, and then multiply by the number of each design the empire owns, instead of processing each ship separately.